File: third/moves.ml
open Printf
open Scanf
open Array

type position = int * int
type node = { position : position; value : int; mutable children : node list }

let print_tree node =
  let rec print_node n =
    printf "Position: %d,%d, Value: %d\n" (fst n.position) (snd n.position) n.value;
    List.iter print_node n.children
  in
  print_node node

let find_neighbors (i, j) n =
  let neighbor_positions = [|-1, -1; -1, 0; -1, 1; 0, -1; 0, 1; 1, -1; 1, 0; 1, 1|] in
  let rows = n in
  let cols = if rows > 0 then n else 0 in
  let is_valid (dr, dc) =
    let new_r, new_c = i + dr, j + dc in
    0 <= new_r && new_r < rows && 0 <= new_c && new_c < cols
  in
  Array.to_list neighbor_positions |> List.filter is_valid |> List.map (fun (dr, dc) -> i + dr, j + dc)
  

let keep_possible neighbors value matrix visited =
  List.filter (fun (i, j) -> value > matrix.(i).(j) && not (List.mem (i, j) visited)) neighbors
  |> List.map (fun (i, j) -> { position = (i, j); value = matrix.(i).(j); children = [] })

let rec expand_tree current_node n matrix visited =
  visited := current_node.position :: !visited;
  let neighbors = find_neighbors current_node.position n in
  let possible_neighbors = keep_possible neighbors current_node.value matrix !visited in
  current_node.children <- possible_neighbors;
  List.iter (fun child -> expand_tree child n matrix (ref (!visited))) current_node.children

let bfs root n =
  let rec bfs_helper queue =
    match queue with
    | [] -> None
    | (current_node, path) :: rest ->
      if current_node.position = (n - 1, n - 1) then Some path
      else begin
        let new_paths = List.map (fun child -> child, path @ [child.position]) current_node.children in
        bfs_helper (rest @ new_paths)
      end
  in
  bfs_helper [(root, [root.position])]

let path_to_moves path =
  match path with
  | None -> None
  | Some p ->
    let rec to_directions acc = function
      | [] -> List.rev acc
      | [ _ ] -> List.rev acc
      | (r1, c1) :: (r2, c2) :: tl ->
        let direction =
          if r2 - r1 = -1 then "N"
          else if r2 - r1 = 1 then "S"
          else "" in
        let direction' =
          if c2 - c1 = -1 then direction ^ "W"
          else if c2 - c1 = 1 then direction ^ "E"
          else direction in
        to_directions (direction' :: acc) ((r2, c2) :: tl)
    in
    Some (Printf.sprintf "[%s]" (String.concat "," (to_directions [] p)))


let () =
  if Array.length Sys.argv <> 2 then begin
    printf "Wrong usage.\n";
    exit 1
  end;

  let filename = Sys.argv.(1) in
  let file = open_in filename in
  let n = int_of_string (input_line file) in
  
  let matrix = Array.make_matrix n n 0 in
  
  for i = 0 to n - 1 do
    let line = input_line file in
    let row = List.map int_of_string (String.split_on_char ' ' line) in
    Array.blit (Array.of_list row) 0 matrix.(i) 0 (List.length row);

  done;
  
  close_in file;
  
  

  let root = { position = (0, 0); value = matrix.(0).(0); children = [] } in
  let visited = ref [(0, 0)] in
  expand_tree root n matrix visited;
  let path = bfs root n in
  let path = path_to_moves path in

  match path with
  | Some p -> printf "%s\n" p
  | None -> printf "IMPOSSIBLE\n"

File: third/moves.py
import sys
import copy
from collections import deque

class Node:
    def __init__(self, position, value):
        self.position = position
        self.value = value
        self.children = []

def print_tree(node):
    if node is None:
        return

    # Print the position and value of the current node
    print(f"Position: {node.position}, Value: {node.value}")
    
    # Recursively print each child
    for child in node.children:
        print_tree(child)

def find_neighbors(position, N):
    i = position[0]
    j = position[1]

    neighbor_positions = [
        (-1, -1), (-1, 0), (-1, 1),
        (0, -1),         (0, 1),
        (1, -1), (1, 0), (1, 1)
    ]
    
    neighbors = []
    rows = N
    cols = N if rows > 0 else 0
    
    for dr, dc in neighbor_positions:
        new_r, new_c = i + dr, j + dc
        # Check if the new position is within the grid bounds
        if 0 <= new_r < rows and 0 <= new_c < cols:
            neighbors.append((new_r, new_c))
    
    return neighbors

def keep_possible(neighbors, value, matrix, visited):
    possible_neighbors = []

    for n in neighbors:
        i = n[0]
        j = n[1]
        if value > matrix[i][j] and (n not in visited):
            node = Node(n, matrix[i][j])
            possible_neighbors.append(node)

    return possible_neighbors

def expand_tree(current_node: Node, N, matrix, visited):
    visited.append(current_node.position)
    neighbors = find_neighbors(current_node.position, N)
    neighbors = keep_possible(neighbors, current_node.value, matrix, visited)
    current_node.children = neighbors
    for child in current_node.children:
        expand_tree(child, N, matrix, copy.deepcopy(visited))

def bfs(root: Node, N):
    queue = deque([(root, [root.position])])

    while queue:
        current_node, path = queue.popleft()
        if current_node.position == (N - 1, N - 1):
            return path
        for child in current_node.children:
            new_path = path + [child.position]  # Extend the current path with the child's position
            queue.append((child, new_path))

    return None

def path_to_moves(path):
    if path is None:
        return None
    else:
        directions = '['
        for i in range(1, len(path)):
            current_step = path[i]
            prev_step = path[i - 1]
            dr = current_step[0] - prev_step[0]
            dc = current_step[1] - prev_step[1]

            # Define directions based on the differences
            direction = ''
            if dr == -1:
                direction += 'N'
            elif dr == 1:
                direction += 'S'

            if dc == 1:
                direction += 'E'
            elif dc == -1:
                direction += 'W'

            directions += direction
            directions += ','

        directions = directions[:-1]
        directions += ']'
        return directions

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Wrong usage.")
        sys.exit()

    filename = sys.argv[1]
    with open(filename, 'r') as file:
        N = int(file.readline().strip())
        
        matrix = []
        
        # Read the next N lines to form the matrix
        for _ in range(N):
            line = file.readline().strip()
            # Split the line into integers and append to the matrix
            row = list(map(int, line.split()))
            matrix.append(row)

    current_pos = (0,0)
    current_value = matrix[0][0]
    root = Node(current_pos, current_value)
    visited = [current_pos]

    expand_tree(root, N, matrix, copy.deepcopy(visited))

    path = bfs(root, N)

    path = path_to_moves(path)

    if path is not None:
        print(path)
    else:
        print("IMPOSSIBLE")



